ocaml/tests/test_psr.ml

```diff
@@ -0,0 +1,340 @@
+          match h.fistpoint with
+          | Some ((Before, Send_new_pool_secret) as fp) ->
+              raise (Fistpoint fp)
+          | Some ((After, Send_new_pool_secret) as fp) ->
+              f () ; raise (Fistpoint fp)
+          | _ ->
+              f ())
+
+    let tell_cleanup_old_pool_secret =
+      iter_host (fun h ->
+          let f () = h.staged_pool_secret <- None in
+          match h.fistpoint with
+          | Some ((Before, Cleanup) as fp) ->
+              raise (Fistpoint fp)
+          | Some ((After, Cleanup) as fp) ->
+              f () ; raise (Fistpoint fp)
+          | _ ->
+              f ())
+
+    let cleanup_master () =
+      let f () =
+        master.member.staged_pool_secret <- None ;
+        master.psr_state.checkpoint <- None ;
+        master.psr_state.pool_secret_backups <- None
+      in
+      match master.member.fistpoint with
+      | Some ((Before, Cleanup) as fp) ->
+          raise (Fistpoint fp)
+      | Some ((After, Cleanup) as fp) ->
+          f () ; raise (Fistpoint fp)
+      | _ ->
+          f ()
+  end
+
+module type PSR = sig
+  val start :
+       string * string
+    -> master:master
+    -> members:member list
+    -> host_id Xapi_psr.r
+end
+
+(* the test implementation has been defined above,
+   and we use this function to access it *)
+let mk_psr master members =
+  let module PSR = Xapi_psr.Make (Impl (struct
+    let master = master
+
+    let members = members
+  end)) in
+  let module PSR = struct
+    include PSR
+
+    let start pool_secrets ~master ~members =
+      start pool_secrets ~master:(Master master)
+        ~members:(List.map (fun m -> Member m) members)
+      |> map_r (function Member m -> m.id | Master m -> m.member.id)
+  end in
+  (module PSR : PSR)
+
+let r' = Alcotest.testable (Fmt.of_to_string string_of_r) ( = )
+
+let check_psr_succeeded r exp_pool_secret master members =
+  let hosts = master.member :: members in
+  Alcotest.check r' "PSR should be successful" (Ok ()) r ;
+  List.iter
+    (fun h ->
+      Alcotest.(
+        check string "new pool secret should be correct" exp_pool_secret
+          h.current_pool_secret))
+    hosts ;
+  List.iter
+    (fun h ->
+      Alcotest.(
+        check (option string) "staged_pool_secret is null" None
+          h.staged_pool_secret))
+    hosts ;
+  let psr_state = master.psr_state in
+  Alcotest.(
+    check (option string) "checkpoint was cleaned up" None psr_state.checkpoint) ;
+  Alcotest.(
+    check
+      (option (pair string string))
+      "pool secret backups were cleaned up" None psr_state.pool_secret_backups)
+
+(* we test almost all combinations of hosts and fistpoints. the only
+   case we don't test is one case of master cleanup failure,
+   since it is slightly different - it has its own unit test (see below).
+ *)
+let almost_all_possible_fists ~num_hosts =
+  let multiply fp_times fp_actions host_ids =
+    List.map
```

Comment:
Maybe next time we'll be able to use `List.concat_map`...

Reply:
:crossed_fingers: